### PR TITLE
Feat(design-tokens): Introduce structure of exported tokens to JS

### PR DESCRIPTION
**Pull request by literat**
### Issue reference

---

### Before submitting the PR, please make sure you do the following

- [ ] Read the [Contributing Guidelines](https://github.com/lmc-eu/spirit-design-system/blob/main/CONTRIBUTING.md).
- [ ] Follow the [PR Title/Commit Message Convention](https://github.com/lmc-eu/spirit-design-system/blob/main/CONTRIBUTING.md#commit-conventions).
- [ ] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

### CODE REVIEW

File: packages/web/src/js/AutoResize.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import BaseComponent from './BaseComponent';
 import { EventHandler, SelectorEngine } from './dom';
 import { enableToggleAutoloader } from './utils';
+import { SpiritConfig } from './utils/Config';
```
Comment (crishpeen) [inline packages/web/src/js/AutoResize.ts:4]: Add Config export to packages/web/src/js/utils/index.ts and import it here on the third row too?

File: packages/web/src/js/BaseComponent.ts [hunk 2]
```diff
@@ -1,17 +1,21 @@
-import { getElement } from './utils/index';
 import InstanceMap from './dom/InstanceMap';
+import Config, { SpiritConfig } from './utils/Config';
+import { getElement } from './utils/index';
```
Comment (crishpeen) [inline packages/web/src/js/BaseComponent.ts:3]: utils/index is required? Couldn't it be just /utils?

File: packages/web/src/js/AutoResize.ts (hunk 1 above)
Comment (crishpeen) [reply packages/web/src/js/AutoResize.ts:4]: This applies to all files.

File: packages/web/src/js/Offcanvas.ts [hunk 3]
```diff
@@ -1,6 +1,9 @@
+// eslint-disable-next-line import/no-unresolved
```
Comment (crishpeen) [inline packages/web/src/js/Offcanvas.ts:1]: :(

File: packages/web/src/js/Offcanvas.ts [hunk 4]
```diff
@@ -11,23 +14,39 @@ const EVENT_SHOWN = `shown${EVENT_KEY}`;
...
+  }
+
+  getBreakpoint() {
+    let breakpoint = OFFCANVAS_BREAKPOINT;
+
+    if (this.config && typeof this.config === 'object' && 'breakpoint' in this.config) {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore Property 'breakpoint' does not exist on type 'object'.
+      breakpoint = this.config.breakpoint as number;
+    } else if (this.element.dataset.spiritBreakpoint) {
```
Comment (crishpeen) [inline packages/web/src/js/Offcanvas.ts:45]: 😨 I didn't know we planned to allow passing from data attribute 😨

File: packages/web/src/js/Offcanvas.ts [hunk 5]
```diff
@@ -11,23 +14,39 @@ const EVENT_SHOWN = `shown${EVENT_KEY}`;
...
     const target = element;
-    super(target);
+    super(target, config);
 
     this.isShown = false;
     this.scrollControl = new ScrollControl(element);
+    this.breakpoint = this.getBreakpoint();
+  }
+
+  getBreakpoint() {
```
Comment (crishpeen) [inline packages/web/src/js/Offcanvas.ts:38]: no CSS variable option? :(

File: packages/web/src/js/dom/__tests__/Manipulator.test.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  beforeAll(() => {
+    fixtureEl = getFixture();
+  });
+
+  afterEach(() => {
+    clearFixture();
+  });
+
+  describe('setDataAttribute', () => {
+    it('should set data attribute prefixed with bs', () => {
```
Comment (crishpeen) [inline packages/web/src/js/dom/__tests__/Manipulator.test.ts:16]: I think some one was copying from other libraries :D

Comment (crishpeen) [reply packages/web/src/js/dom/__tests__/Manipulator.test.ts:16]: no bs prefix :)

File: packages/web/src/js/utils/Config.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
+import Manipulator from '../dom/Manipulator';
+import { isElement } from './index';
+
+export type SpiritConfig = Record<string, unknown> | null;
+
+// Shout-out Angus Croll (https://goo.gl/pxwQGp)
```
Comment (crishpeen) [inline packages/web/src/js/utils/Config.ts:6]: oh wow, more than 12 years old article, but still relevant I guess :)

File: packages/web/src/js/utils/__tests__/Config.test.ts [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        testBool: true,
+        testString: 'foo',
+        testString1: 'foo',
+        testInt: 7,
+      });
+
+      const configResult = DummyConfigClass.mergeConfigObj({}, fixtureEl.querySelector('#test'));
+
+      expect(configResult?.testBool).toBe(false);
+      // expect(configResult?.testString).toBe('foo');
```
Comment (crishpeen) [inline packages/web/src/js/utils/__tests__/Config.test.ts:55]: Do we want to keep this?

File: packages/web/src/js/utils/__tests__/Config.test.ts [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+      const configResult = DummyConfigClass.mergeConfigObj(
+        {
+          testString1: 'test',
+          testInt: 3,
+        },
+        fixtureEl.querySelector('#test'),
+      );
+
+      expect(configResult?.testBool).toBe(false);
+      // expect(configResult?.testString).toBe('foo');
```
Comment (crishpeen) [inline packages/web/src/js/utils/__tests__/Config.test.ts:80]: and this?

File: packages/web/src/js/utils/__tests__/Config.test.ts [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+      jest.spyOn(DummyConfigClass, 'Default', 'get').mockReturnValue({
+        testInt: 7,
+        testInt2: 79,
+      });
+
+      const configResult = DummyConfigClass.mergeConfigObj({}, fixtureEl.querySelector('#test'));
+
+      expect(configResult?.testInt).toBe(8);
+      // expect(configResult?.testInt2).toBe(79);
```
Comment (crishpeen) [inline packages/web/src/js/utils/__tests__/Config.test.ts:122]: and this?

**Review by crishpeen — CHANGES_REQUESTED** [review]
I feel like this is several PRs merged into one. It is quite hard to navigate through this PR, especially without any README. I assume we need some at least for the Offcanvas plugin, because now only you know how to set the breakpoint. So please document this better.

Also, I am a bit afraid if it won't break anything, as you modify a lot of files and packages, but hopefully it won't. Also I didn't find the CSS variable implementation?

File: packages/web/src/js/dom/__tests__/Manipulator.test.ts (hunk 6 above)
Comment (literat) [reply packages/web/src/js/dom/__tests__/Manipulator.test.ts]: Fixed.

File: packages/web/src/js/AutoResize.ts (hunk 1 above)
Comment (literat) [reply packages/web/src/js/AutoResize.ts]: Fixed.

File: packages/web/src/js/Offcanvas.ts (hunk 3 above)
Comment (literat) [reply packages/web/src/js/Offcanvas.ts]: Icons need to be built first.

File: packages/web/src/js/Offcanvas.ts (hunk 5 above)
Comment (literat) [reply packages/web/src/js/Offcanvas.ts:38]: Rebasing 🤯

Comment (literat) [reply packages/web/src/js/Offcanvas.ts:38]: Fixed.

File: packages/web/src/js/utils/__tests__/Config.test.ts (hunk 8 above)
Comment (literat) [reply packages/web/src/js/utils/__tests__/Config.test.ts:55]: I am trying to figure out the mocking of the static properties.

Comment (literat) [reply packages/web/src/js/utils/__tests__/Config.test.ts]: Fixed.

File: packages/web/src/js/utils/__tests__/Config.test.ts (hunk 9 above)
Comment (literat) [reply packages/web/src/js/utils/__tests__/Config.test.ts]: Fixed.

File: packages/web/src/js/utils/__tests__/Config.test.ts (hunk 10 above)
Comment (literat) [reply packages/web/src/js/utils/__tests__/Config.test.ts]: Fixed.

**Comment by literat** [comment]
> I feel like this is several PRs merged into one. It is quite hard to navigate through this PR, especially without any README. I assume we need some at least for the Offcanvas plugin, because now only you know how to set the breakpoint. So please document this better.

Try to go commit-by-commit. The build process is a little bit complicated because design tokens must be built first.

> Also, I am a bit afraid if it won't break anything, as you modify a lot of files and packages, but hopefully it won't. Also I didn't find the CSS variable implementation?

This was a problem with the rebase I have made. The CSS variable was added by me and also I have introduced the Offcanvas Readme.

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+You can create an offcanvas instance with the constructor, for example:
+
+```js
+const customOffcanvas = document.getElementById('customOffcanvas');
+const spiritOffcanvas = new Offcanvas(customOffcanvas);
+```
+
+| Method              | Description                                                                                                                                                                                              |
+| ------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| toggle              | Toggles an Offcanvas element to shown or hidden. Returns to the caller before the Offcanvas element has actually been shown or hidden (i.e. before the show.offcanvas or hidden.offcanvas event occurs). |
```
Comment (crishpeen) [inline packages/web/src/scss/components/Offcanvas/README.md:57]: Place add backticks to all items in Method column and in the Event type too.

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
+# Offcanvas
+
+Offcanvas is a sidebar component that can be toggled via JavaScript to appear from the left, right, or bottom edge of the viewport.
+Buttons or anchors are used as triggers that are attached to specific elements you toggle, and data attributes are used to invoke JavaScript plugin.
+
+## JavaScript Plugin
```
Comment (crishpeen) [inline packages/web/src/scss/components/Offcanvas/README.md:6]: Link to the main readme for how to include JavaScript plugins?

File: exporters/js/src/js/printers/unit.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
+export function printUnit(value: number, unit: string): string {
+  let result = value.toString();
+  if (+value !== 0) {
+    if (unit === 'Pixels') {
```
Comment (pavelklibani) [inline exporters/js/src/js/printers/unit.ts:4]: Why is it with capital "P"? 
Maybe lowercase `unit` and then compare to "pixels".

Comment (crishpeen) [reply exporters/js/src/js/printers/unit.ts:4]: Supernova returns pixels as Pixels 🤷 It is a good idea to lowercase it and then do the compare, but IMHO it is not really important to do now :) @literat you can improve this (and the same code is in the scss exporter too) if you want to

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+const offcanvasElementList = [].slice.call(document.querySelectorAll('.offcanvas'));
+const offcanvasList = offcanvasElementList.map(function (offcanvasElement) {
+  return new Offcanvas(offcanvasElement);
+});
+```
+
+### Options
+
+Options can be passed via data attributes or JavaScript.
+For data attributes, append the option name to data-spirit-, as in data-spirit-breakpointDesktop="".
```
[the remark this answers is not in the document]
Comment (crishpeen) [reply packages/web/src/scss/components/Offcanvas/README.md:36]: Other multiword breakpoints use the kebab case. Can we please have this as kebab-case too?
eg. `data-spirit-close-on-backdrop-click`

```suggestion
For data attributes, append the option name to data-spirit-, as in data-spirit-breakpoint-desktop="".
```

File: packages/web/src/js/Modal.ts [hunk 15]
```diff
@@ -1,7 +1,7 @@
 import BaseComponent from './BaseComponent';
 import EventHandler from './dom/EventHandler';
 import SelectorEngine from './dom/SelectorEngine';
-import { enableToggleTrigger, ScrollControl } from './utils';
+import { enableToggleTrigger, ScrollControl, SpiritConfig } from './utils';
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web/src/js/Modal.ts:4]: This is sorted by my IDE. I can switch this off. Or we should discuss if this is a big topic for us.

File: packages/web/src/scss/components/Offcanvas/README.md (hunk 12 above)
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md:6]: Linked.

File: packages/web/src/scss/components/Offcanvas/README.md (hunk 11 above)
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md]: Added.

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+| Name              | Type   | Default | Description                                                       |
+| ----------------- | ------ | ------- | ----------------------------------------------------------------- |
+| breakpointDesktop | string | 1280 \* | Desktop breakpoint on which opened Offcanvas automatically closes |
+
+\*: Default value of the design token `others.breakpoints.desktop`. Can be also changed via `--spirit-breakpoint-desktop` CSS custom property.
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md]: Fixed.

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+```
+
+### Options
+
+Options can be passed via data attributes or JavaScript.
+For data attributes, append the option name to data-spirit-, as in data-spirit-breakpointDesktop="".
+
+| Name              | Type   | Default | Description                                                       |
+| ----------------- | ------ | ------- | ----------------------------------------------------------------- |
+| breakpointDesktop | string | 1280 \* | Desktop breakpoint on which opened Offcanvas automatically closes |
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md]: Fixed.

File: packages/web/src/scss/components/Offcanvas/README.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md]: Fixed.

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+```html
+<script src="node_modules/@lmc-eu/spirit-web/js/cjs/spirit-web.min.js" async></script>
+```
+
+## Usage
+
+### Via data attributes
+
+Add data-spirit-toggle="offcanvas" and a data-spirit-target or href to the element to automatically assign control of one Offcanvas element.
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md]: Example added.

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 19]
```diff
@@ -0,0 +1,79 @@
+# Offcanvas
+
+Offcanvas is a sidebar component that can be toggled via JavaScript to appear from the left, right, or bottom edge of the viewport.
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md]: Directions removed.

File: packages/web/src/js/Offcanvas.ts [hunk 20]
```diff
@@ -11,23 +12,52 @@ const EVENT_SHOWN = `shown${EVENT_KEY}`;
...
+};
+
+const DefaultType = {
+  breakpointDesktop: 'number',
+};
+
 class Offcanvas extends BaseComponent {
   isShown: boolean;
   scrollControl: ScrollControl;
+  breakpoint: number;
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web/src/js/Offcanvas.ts]: Sorted.

**Review by pavelklibani — APPROVED** [review]

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+```
+
+### Options
+
+Options can be passed via data attributes or JavaScript.
+For data attributes, append the option name to `data-spirit-`, as in `data-spirit-breakpoint-desktop="1280"`.
+
+| Name                | Type     | Default | Description                                                       |
+| ------------------- | -------- | ------- | ----------------------------------------------------------------- |
+| `breakpointDesktop` | `string` | 1280 \* | Desktop breakpoint on which opened Offcanvas automatically closes |
```
Comment (crishpeen) [inline packages/web/src/scss/components/Offcanvas/README.md:58]:
```suggestion
| `breakpoint-desktop` | `string` | 1280 \* | Desktop breakpoint on which opened Offcanvas automatically closes |
```
?

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+
+\*: Default value of the design token `other.breakpoint.desktop`. Can be also changed via `--spirit-breakpoint-desktop` CSS custom property.
```
Comment (crishpeen) [inline packages/web/src/scss/components/Offcanvas/README.md:60]:
```suggestion
\*: Default value of the design token `other.breakpoint.desktop`. Can be also changed via `--spirit-breakpoint-desktop` CSS custom property. This CSS property is pre-defined in our foundation CSS.
```

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 23]
```diff
@@ -0,0 +1,99 @@
+# Offcanvas
+
+Offcanvas is a sidebar component that can be toggled via JavaScript to appear from the right edge of the viewport.
+Buttons or anchors are used as triggers that are attached to specific elements you toggle, and data attributes are used to invoke JavaScript plugin.
+
+## JavaScript Plugin
+
+For full functionality, you need to [provide Spirit JavaScript][js-plugin-readme], which will handle
+toggling of the Header component:
```
Comment (crishpeen) [inline packages/web/src/scss/components/Offcanvas/README.md:9]: Header? 

It is first time mentioned :)

File: packages/web/src/scss/components/Offcanvas/README.md [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+| Event type         | Description                                                                                                                  |
+| ------------------ | ---------------------------------------------------------------------------------------------------------------------------- |
+| `show.offcanvas`   | This event fires immediately when the show instance method is called.                                                        |
+| `shown.offcanvas`  | This event is fired when an offcanvas element has been made visible to the user (will wait for CSS transitions to complete). |
+| `hide.offcanvas`   | This event is fired immediately when the hide method has been called.                                                        |
+| `hidden.offcanvas` | This event is fired when an offcanvas element has been hidden from the user (will wait for CSS transitions to complete)      |
+
+```js
+const myOffcanvas = document.getElementById('customOffcanvas');
+customOffcanvas.addEventListener('hidden.offcanvas', function () {
```
Comment (crishpeen) [inline packages/web/src/scss/components/Offcanvas/README.md:94]:
```suggestion
myOffcanvas.addEventListener('hidden.offcanvas', function () {
```

**Review by crishpeen — APPROVED** [review]

**Review by crishpeen — COMMENTED** [review]
few small fixes and it is hopefully ready to merge

File: packages/web/src/scss/components/Offcanvas/README.md (hunk 24 above)
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md]: Fixed.

File: packages/web/src/scss/components/Offcanvas/README.md (hunk 23 above)
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md]: Fixed.

File: packages/web/src/scss/components/Offcanvas/README.md (hunk 22 above)
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md:60]: Fixed.

File: packages/web/src/scss/components/Offcanvas/README.md (hunk 21 above)
Comment (literat) [reply packages/web/src/scss/components/Offcanvas/README.md]: Fixed.

**Pull request merged**